storage/local/storage.go

```diff
@@ -559,25 +573,29 @@ func (s *memorySeriesStorage) Append(sample *clientmodel.Sample) {
 	s.incNumChunksToPersist(completedChunksCount)
 }
 
-func (s *memorySeriesStorage) getOrCreateSeries(fp clientmodel.Fingerprint, m clientmodel.Metric) *memorySeries {
+func (s *memorySeriesStorage) getOrCreateSeries(fp clientmodel.Fingerprint, m clientmodel.Metric) (series *memorySeries, lastAppendTime clientmodel.Timestamp) {
 	series, ok := s.fpToSeries.get(fp)
 	if !ok {
-		unarchived, firstTime, err := s.persistence.unarchiveMetric(fp)
+		unarchived, firstTime, lastTime, err := s.persistence.unarchiveMetric(fp)
 		if err != nil {
 			log.Errorf("Error unarchiving fingerprint %v: %v", fp, err)
 		}
 		if unarchived {
 			s.seriesOps.WithLabelValues(unarchive).Inc()
+			lastAppendTime = lastTime
 		} else {
 			// This was a genuinely new series, so index the metric.
 			s.persistence.indexMetric(fp, m)
 			s.seriesOps.WithLabelValues(create).Inc()
+			lastAppendTime = clientmodel.Earliest
 		}
 		series = newMemorySeries(m, !unarchived, firstTime)
 		s.fpToSeries.put(fp, series)
 		s.numSeries.Inc()
+	} else {
+		lastAppendTime = series.head().lastTime()
```

Comment:
Plus the allocation churn for the chunk iterator... up to 360,000 times per second... ;)